File: js/src/accounts/PassBook.ts
import {
  Borsh,
  StringPublicKey,
  AnyPublicKey,
  ERROR_INVALID_OWNER,
  Account,
} from '@metaplex-foundation/mpl-core';
import bs58 from 'bs58';
import { AccountInfo, Connection, PublicKey } from '@solana/web3.js';
import { PassBookProgram } from '../PassBookProgram';
import { AccountKey, PassState, PassType } from './constants';

type Args = {
  key: AccountKey;
  mint: StringPublicKey;
  authority: StringPublicKey;
  name: string;
  description: string;
  uri: string;
  mutable: boolean;
  passType: PassType;
  validityPeriod: number | null;
  collectionMint: StringPublicKey | null;
  timeValidationAuthority: StringPublicKey | null;
  passState: PassState;
};

export class PassBookData extends Borsh.Data<Args> {
  static readonly SCHEMA = PassBookData.struct([
    ['key', 'u8'],
    ['authority', 'pubkeyAsString'],
    ['mint', 'pubkeyAsString'],
    ['name', 'string'],
    ['description', 'string'],
    ['uri', 'string'],
    ['mutable', 'u8'],
    ['passType', 'u8'],
    ['validityPeriod', { kind: 'option', type: 'u32' }],
    ['collectionMint', { kind: 'option', type: 'pubkeyAsString' }],
    ['timeValidationAuthority', { kind: 'option', type: 'pubkeyAsString' }],
    ['passState', 'u8'],
  ]);
  key: AccountKey;
  mint: StringPublicKey;
  authority: StringPublicKey;
  name: string;
  description: string;
  uri: string;
  mutable: boolean;
  passType: PassType;
  validityPeriod?: number;
  collectionMint?: StringPublicKey;
  timeValidationAuthority?: StringPublicKey;
  passState: PassState;

  constructor(args: Args) {
    super(args);
    const REPLACE = new RegExp('\u0000', 'g');
    this.key = AccountKey.PassBook;
    this.name = args.name.replace(REPLACE, '');
    this.description = args.description.replace(REPLACE, '');
    this.uri = args.uri.replace(REPLACE, '');
  }
}

export class PassBook extends Account<PassBookData> {
  constructor(pubkey: AnyPublicKey, info: AccountInfo<Buffer>) {
    super(pubkey, info);
    this.data = PassBookData.deserialize(this.info.data);
    if (!this.assertOwner(PassBookProgram.PUBKEY)) {
      throw ERROR_INVALID_OWNER();
    }
  }

  static async getPDA(mint: AnyPublicKey) {
    return PassBookProgram.findProgramAddress([
      Buffer.from(PassBookProgram.PREFIX),
      PassBookProgram.PUBKEY.toBuffer(),
      new PublicKey(mint).toBuffer(),
    ]);
  }

  static async findMany(
    connection: Connection,
    filters: {
      mint?: AnyPublicKey;
      authority?: AnyPublicKey;
    } = {},
  ) {
    const baseFilters = [
      // Filter for PassBook by key
      {
        memcmp: {
          offset: 0,
          bytes: bs58.encode(Buffer.from([AccountKey.PassBook])),
        },
      },
      // Filter for assigned to authority
      filters.authority && {
        memcmp: {
          offset: 1,
          bytes: new PublicKey(filters.authority).toBase58(),
        },
      },
      // Filter for assigned to mint
      filters.mint && {
        memcmp: {
          offset: 33,
          bytes: new PublicKey(filters.mint).toBase58(),
        },
      },
    ].filter(Boolean);

    return (await PassBookProgram.getProgramAccounts(connection, { filters: baseFilters })).map(
      (account) => PassBook.from(account),
    );
  }

  static async findByMint(connection: Connection, mint: AnyPublicKey): Promise<PassBook> {
    const pda = await PassBook.getPDA(mint);
    return PassBook.load(connection, pda);
  }

  static async findByAuthority(
    connection: Connection,
    authority: AnyPublicKey,
  ): Promise<Account<PassBook>[]> {
    return await PassBook.findMany(connection, {
      authority,
    });
  }
}

File: dart/lib/accounts/passbook.dart
import 'dart:typed_data';

import 'package:passbook/accounts/constants.dart';
import 'package:passbook/utils/endian.dart';
import 'package:passbook/utils/struct_reader.dart';
import 'package:solana/base58.dart';

class PassBook {
  const PassBook(
      {required this.key,
      required this.name,
      required this.description,
      required this.uri,
      required this.authority,
      required this.mint,
      required this.mutable,
      required this.passState,
      required this.durationType,
      required this.duration,
      required this.totalPasses,
      this.maxSupply});

  factory PassBook.fromBinary(List<int> sourceBytes) {
    final bytes = Int8List.fromList(sourceBytes);
    final reader = StructReader(bytes.buffer)..skip(1);
    final authority = base58encode(reader.nextBytes(32));
    final mint = base58encode(reader.nextBytes(32));
    final name = reader.nextString();
    final description = reader.nextString();
    final uri = reader.nextString();
    final mutable = reader.nextBytes(1).first == 1;
    final passState = PassStateExtension.fromId(reader.nextBytes(1).first);
    final durationType =
        DurationTypeExtension.fromId(reader.nextBytes(1).first);
    final BigInt duration = decodeBigInt(reader.nextBytes(8), Endian.little);
    final BigInt totalPasses = decodeBigInt(reader.nextBytes(8), Endian.little);
    final hasMaxSupply = reader.nextBytes(1).first == 1;
    final BigInt? maxSupply =
        hasMaxSupply ? decodeBigInt(reader.nextBytes(8), Endian.little) : null;
    return PassBook(
        key: AccountKey.passBook,
        name: name,
        description: description,
        uri: uri,
        authority: authority,
        mint: mint,
        mutable: mutable,
        passState: passState,
        durationType: durationType,
        duration: duration,
        totalPasses: totalPasses,
        maxSupply: maxSupply);
  }

  final AccountKey key;
  final String name;
  final String description;
  final String uri;
  final String authority;
  final String mint;
  final bool mutable;
  final PassState passState;
  final DurationType durationType;
  final BigInt duration;
  final BigInt totalPasses;
  final BigInt? maxSupply;
}
